// @letta-ai/letta-client/serialization/types/AppAuthScheme.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppAuthScheme = void 0;
const core = __importStar(require("../../core"));
const AppAuthSchemeAuthMode_1 = require("./AppAuthSchemeAuthMode");
const AuthSchemeField_1 = require("./AuthSchemeField");
exports.AppAuthScheme = core.serialization.object({
    schemeName: core.serialization.property("scheme_name", core.serialization.string()),
    authMode: core.serialization.property("auth_mode", AppAuthSchemeAuthMode_1.AppAuthSchemeAuthMode),
    fields: core.serialization.list(AuthSchemeField_1.AuthSchemeField),
    proxy: core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional(),
    authorizationUrl: core.serialization.property("authorization_url", core.serialization.string().optional()),
    tokenUrl: core.serialization.property("token_url", core.serialization.string().optional()),
    defaultScopes: core.serialization.property("default_scopes", core.serialization.list(core.serialization.unknown()).optional()),
    tokenResponseMetadata: core.serialization.property("token_response_metadata", core.serialization.list(core.serialization.unknown()).optional()),
    clientId: core.serialization.property("client_id", core.serialization.string().optional()),
    clientSecret: core.serialization.property("client_secret", core.serialization.string().optional()),
});
