// @letta-ai/letta-client/serialization/types/AppModel.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppModel = void 0;
const core = __importStar(require("../../core"));
const AppAuthScheme_1 = require("./AppAuthScheme");
exports.AppModel = core.serialization.object({
    name: core.serialization.string(),
    key: core.serialization.string(),
    appId: core.serialization.string(),
    description: core.serialization.string(),
    categories: core.serialization.list(core.serialization.string()),
    meta: core.serialization.record(core.serialization.string(), core.serialization.unknown()),
    logo: core.serialization.string().optional(),
    docs: core.serialization.string().optional(),
    group: core.serialization.string().optional(),
    status: core.serialization.string().optional(),
    enabled: core.serialization.boolean().optional(),
    noAuth: core.serialization.property("no_auth", core.serialization.boolean().optional()),
    authSchemes: core.serialization.property("auth_schemes", core.serialization.list(AppAuthScheme_1.AppAuthScheme).optional()),
    testConnectors: core.serialization
        .list(core.serialization.record(core.serialization.string(), core.serialization.unknown()))
        .optional(),
    documentationDocText: core.serialization.property("documentation_doc_text", core.serialization.string().optional()),
    configurationDocsText: core.serialization.property("configuration_docs_text", core.serialization.string().optional()),
});
